bgpd/bgp_nb_config.c

```diff
@@ -67,12 +202,63 @@ int routing_control_plane_protocols_control_plane_protocol_bgp_global_destroy(
 int routing_control_plane_protocols_control_plane_protocol_bgp_global_local_as_modify(
 	struct nb_cb_modify_args *args)
 {
+	const struct lyd_node *vrf_dnode;
+	const char *vrf_name;
+	struct bgp *bgp;
+	const char *name = NULL;
+	as_t as;
+	enum bgp_instance_type inst_type;
+	int ret;
+	bool is_view_inst = false;
+
 	switch (args->event) {
 	case NB_EV_VALIDATE:
+		inst_type = BGP_INSTANCE_TYPE_DEFAULT;
+
+		vrf_dnode = yang_dnode_get_parent(args->dnode,
+						  "control-plane-protocol");
+		vrf_name = yang_dnode_get_string(vrf_dnode, "./vrf");
+
+		as = yang_dnode_get_uint32(args->dnode, NULL);
+
+		if (strmatch(vrf_name, VRF_DEFAULT_NAME)) {
+			name = NULL;
+		} else {
+			name = vrf_name;
+			inst_type = BGP_INSTANCE_TYPE_VRF;
+		}
+
+		is_view_inst = yang_dnode_get_bool(args->dnode,
+						   "../instance-type-view");
+		if (is_view_inst)
+			inst_type = BGP_INSTANCE_TYPE_VIEW;
+
+		ret = bgp_lookup_by_as_name_type(&bgp, &as, name, inst_type);
+		if (ret == BGP_ERR_INSTANCE_MISMATCH) {
+			snprintf(
+				args->errmsg, args->errmsg_len,
+				"BGP instance name and AS number mismatch\nBGP instance is already running; AS is %u",
+				as);
+
+			return NB_ERR_VALIDATION;
+		}
+
+		break;
 	case NB_EV_PREPARE:
 	case NB_EV_ABORT:
+		return NB_OK;
 	case NB_EV_APPLY:
-		/* TODO: implement me. */
+		/* NOTE: handled in bgp_global_create callback, the as change
+		 * will be rejected in validate phase.
+		 */
+		as = yang_dnode_get_uint32(args->dnode, NULL);
+		bgp = nb_running_get_entry(args->dnode, NULL, true);
+		if (bgp->as != as) {
```

Comment:
I believe this callback can be simplified a bunch. The following code during the validation phase should be enough in theory:
```
bgp = nb_running_get_entry_non_rec(args->dnode, NULL, false);
if (bgp && bgp->as != as) {
    ...
}
```
That's simpler than looking up a BGP instance by its name, AS and type (fetching all that data requires a lot of code).